### PR TITLE
[adobepass] Add Suddenlink MSO

### PR DESCRIPTION
**Pull request by CplPwnies**
Addresses issue #2704 to add Suddenlink MSO authentication

## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

Addresses issue #2704 to add Suddenlink MSO for authentication

### CODE REVIEW

File: yt_dlp/extractor/adobepass.py [hunk 1]
```diff
@@ -1635,6 +1640,56 @@ def extract_redirect_url(html, url=None, fatal=False):
...
+                    last_bookend_page, urlh = self._download_webpage_handle(
+                        provider_refresh_redirect_url, video_id,
+                        'Downloading Auth Association Redirect Page')
+
+                    hidden_data = self._hidden_inputs(last_bookend_page)
+                    hidden_data['history_val'] = 3
+
+                    mvpd_confirm_page_res = self._download_webpage_handle(
+                        urlh.geturl(), video_id, 'Sending Final Bookend.',
+                        query=hidden_data)
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1691]:
```suggestion
                    provider_login_redirect_page, urlh = self._download_webpage_handle(
                        urlh.geturl(), video_id, 'Sending First Bookend',
                        query=hidden_data)

                    provider_tryauth_url = self._html_search_regex(
                        r'url:\s*[\'"]([^\'"]+)', provider_login_redirect_page, 'ajaxurl')

                    provider_tryauth_page, urlh = self._download_webpage_handle(
                        provider_tryauth_url, video_id, 'Submitting TryAuth',
                        query=hidden_data)

                    provider_login_url = f'https://authorize.suddenlink.net/saml/module.php/authSynacor/login.php?AuthState={provider_tryauth_page}'
                    provider_login_page_res = self._download_webpage_handle(
                        provider_login_url, video_id, 'Getting Login Page.',
                        query=hidden_data)

                    provider_association_redirect, urlh = post_form(
                        provider_login_page_res, 'Logging in', {
                            mso_info['username_field']: username,
                            mso_info['password_field']: password
                        })

                    provider_refresh_redirect_url = extract_redirect_url(
                        provider_association_redirect, url=urlh.geturl())

                    last_bookend_page, urlh = self._download_webpage_handle(
                        provider_refresh_redirect_url, video_id,
                        'Downloading Auth Association Redirect Page')

                    hidden_data = self._hidden_inputs(last_bookend_page)
                    hidden_data['history_val'] = 3

                    mvpd_confirm_page_res = self._download_webpage_handle(
                        urlh.geturl(), video_id, 'Sending Final Bookend',
                        query=hidden_data)
```

File: yt_dlp/extractor/adobepass.py [hunk 2]
```diff
@@ -1635,6 +1640,56 @@ def extract_redirect_url(html, url=None, fatal=False):
...
+                        r'url:\s*[\'"]([^\'"]+)', provider_login_redirect_page, 'ajaxurl')
+
+                    provider_tryauth_page, urlh = self._download_webpage_handle(
+                        provider_tryauth_url, video_id, 'Submitting TryAuth',
+                        query=hidden_data)
+
+                    provider_login_url = f'https://authorize.suddenlink.net/saml/module.php/authSynacor/login.php?AuthState={provider_tryauth_page}'
+                    provider_login_page_res = self._download_webpage_handle(
+                        provider_login_url, video_id, 'Getting Login Page.',
+                        query=hidden_data)
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1671]:
```suggestion
                    provider_login_page_res = self._download_webpage_handle(
                        f'https://authorize.suddenlink.net/saml/module.php/authSynacor/login.php?AuthState={provider_tryauth_page}',
                        video_id, 'Getting Login Page.',
                        query=hidden_data)
```

File: yt_dlp/extractor/adobepass.py [hunk 3]
```diff
@@ -1635,6 +1640,56 @@ def extract_redirect_url(html, url=None, fatal=False):
                         urlh.geturl(), video_id, 'Sending final bookend',
                         query=hidden_data)
 
+                    post_form(mvpd_confirm_page_res, 'Confirming Login')
+                elif mso_id == 'Suddenlink':
+                    # Suddenlink is similar to SlingTV in using a tab history count and a meta refresh,
+                    # but they also do a dynmaic redirect using javascript that has to be followed as well
+                    provider_redirect_page, urlh = provider_redirect_page_res
+
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1648]: unused?
```suggestion
```

File: yt_dlp/extractor/adobepass.py [hunk 4]
```diff
@@ -1635,6 +1640,56 @@ def extract_redirect_url(html, url=None, fatal=False):
...
+                    provider_auth_redirect_res = post_form(
+                        provider_redirect_page_res, 'Pressing Continue...')
+
+                    first_bookend_page, urlh = provider_auth_redirect_res
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1652]:
```suggestion
                    first_bookend_page, urlh = post_form(
                        provider_redirect_page_res, 'Pressing Continue...')
```

File: yt_dlp/extractor/adobepass.py [hunk 5]
```diff
@@ -1635,6 +1640,56 @@ def extract_redirect_url(html, url=None, fatal=False):
...
+
+                    hidden_data = self._hidden_inputs(first_bookend_page)
+                    hidden_data['history_val'] = 1
+
+                    provider_login_redirect_page, urlh = self._download_webpage_handle(
+                        urlh.geturl(), video_id, 'Sending First Bookend',
+                        query=hidden_data)
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1659]:
```suggestion
                    provider_login_redirect_page, _ = self._download_webpage_handle(
                        urlh.geturl(), video_id, 'Sending First Bookend',
                        query=hidden_data)
```
urlh is unused?

File: yt_dlp/extractor/adobepass.py [hunk 6]
```diff
@@ -1635,6 +1640,56 @@ def extract_redirect_url(html, url=None, fatal=False):
...
+
+                    provider_tryauth_url = self._html_search_regex(
+                        r'url:\s*[\'"]([^\'"]+)', provider_login_redirect_page, 'ajaxurl')
+
+                    provider_tryauth_page, urlh = self._download_webpage_handle(
+                        provider_tryauth_url, video_id, 'Submitting TryAuth',
+                        query=hidden_data)
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1666]:
```suggestion
                    provider_tryauth_page, _ = self._download_webpage_handle(
                        provider_tryauth_url, video_id, 'Submitting TryAuth',
                        query=hidden_data)
```
again

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/adobepass.py (hunk 3 above)
Comment (CplPwnies) [reply yt_dlp/extractor/adobepass.py:1648]: After reviewing this myself, `provider_redirect_page` var is unused. I was mostly keying off other examples in the code. Can be removed.

File: yt_dlp/extractor/adobepass.py [hunk 7]
```diff
@@ -1635,6 +1640,52 @@ def extract_redirect_url(html, url=None, fatal=False):
...
+                    provider_login_redirect_page, _ = self._download_webpage_handle(
+                        urlh.geturl(), video_id, 'Sending First Bookend',
+                        query=hidden_data)
+
+                    provider_tryauth_url = self._html_search_regex(
+                        r'url:\s*[\'"]([^\'"]+)', provider_login_redirect_page, 'ajaxurl')
+
+                    provider_tryauth_page, _ = self._download_webpage_handle(
+                        provider_tryauth_url, video_id, 'Submitting TryAuth',
+                        query=hidden_data)
```
Comment (pukkandan) [inline yt_dlp/extractor/adobepass.py:1662]:
```suggestion
                    provider_login_redirect_page = self._download_webpage(
                        urlh.geturl(), video_id, 'Sending First Bookend',
                        query=hidden_data)

                    provider_tryauth_url = self._html_search_regex(
                        r'url:\s*[\'"]([^\'"]+)', provider_login_redirect_page, 'ajaxurl')

                    provider_tryauth_page = self._download_webpage(
                        provider_tryauth_url, video_id, 'Submitting TryAuth',
                        query=hidden_data)
```

**Pull request merged**